File: src/PcIfy.Server/Services/AuthService.cs
using PcIfy.Server.Helpers;
using PcIfy.Server.Models;
using PcIfy.Server.Services.Interfaces;

namespace PcIfy.Server.Services;

public class AuthService : IAuthService
{
    private readonly AppSettings _settings;

    public AuthService(AppSettings settings) => _settings = settings;

    public bool ValidateCredentials(string username, string password)
    {
        var user = _settings.Users.FirstOrDefault(u =>
            string.Equals(u.Username, username, StringComparison.OrdinalIgnoreCase));

        return user is not null && PasswordHasher.Verify(password, user.PasswordHash);
    }

    public string GenerateToken(string username) =>
        JwtHelper.GenerateToken(username, _settings.JwtSecret, _settings.TokenExpiryHours);
}

File: src/PcIfy.Server.Flutter/lib/services/auth_service.dart
import 'package:bcrypt/bcrypt.dart';
import 'package:dart_jsonwebtoken/dart_jsonwebtoken.dart';
import '../core/models/app_settings.dart';

class AuthService {
  final AppSettings settings;

  AuthService(this.settings);

  bool validateCredentials(String username, String password) {
    final user = settings.users
        .where((u) => u.username.toLowerCase() == username.toLowerCase())
        .firstOrNull;
    if (user == null) return false;
    return BCrypt.checkpw(password, user.passwordHash);
  }

  String generateToken(String username) {
    final jwt = JWT(
      {'sub': username, 'name': username},
      issuer: 'pcify-server',
      audience: Audience(['pcify-client']),
    );
    return jwt.sign(
      SecretKey(settings.jwtSecret),
      algorithm: JWTAlgorithm.HS256,
      expiresIn: Duration(hours: settings.tokenExpiryHours),
    );
  }

  /// Verifies a token and returns the username claim, or null if invalid.
  String? verifyToken(String token) {
    try {
      final jwt = JWT.verify(
        token,
        SecretKey(settings.jwtSecret),
        checkHeaderType: false,
        audience: Audience(['pcify-client']),
        issuer: 'pcify-server',
      );
      return jwt.payload['name'] as String?;
    } catch (_) {
      return null;
    }
  }

  static String hashPassword(String password) =>
      BCrypt.hashpw(password, BCrypt.gensalt(logRounds: 12));
}
